lib/puppet/pops/validation/checker4_0.rb

```diff
@@ -530,6 +532,102 @@ def internal_check_future_reserved_word(o, name)
+      acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    end
+  end
+
+  # @api private
+  class Puppet::Util::FileNamespaceAdapter < Puppet::Pops::Adaptable::Adapter
+    attr_accessor :file_to_namespace
+  end
+
+  def namespace_for_file(file)
+    env = Puppet.lookup(:current_environment)
+    return :no_namespace if env.nil?
+
+    Puppet::Util::FileNamespaceAdapter.adapt(env) do |adapter|
+      adapter.file_to_namespace ||= {}
+
+      lc_file = file.downcase
+      file_namespace = adapter.file_to_namespace[lc_file]
+      return file_namespace unless file_namespace.nil? # No cache entry, so we do the calculation
+
+      path = Pathname.new(file)
+
+      return adapter.file_to_namespace[lc_file] = :no_namespace if path.extname != ".pp"
+
+      #We want an expanded path, but lowercase, so...
+      path = Pathname.new(path.expand_path.to_s.downcase)
+
+      return adapter.file_to_namespace[lc_file] = :no_namespace if initial_manifest?(path, env.manifest)
+
+      #All auto-loaded files from modules come from a module search path dir
+      relative_path = get_module_relative_path(path, env.full_modulepath)
+
+      return adapter.file_to_namespace[lc_file] = :no_namespace if relative_path == :no_path
+
+      #If a file comes from a module, but isn't in the right place, always error
+      names = dir_to_names(relative_path)
+
+      return adapter.file_to_namespace[lc_file] = (names == :bad_module_file ? :bad_module_file : names.join("::").freeze)
+    end
+  end
+
+  def initial_manifest?(path, manifest_setting)
+    return false if manifest_setting.nil? || manifest_setting == :no_manifest
+
+    string_path = path.to_s
+
+    string_path == manifest_setting || string_path.start_with?(manifest_setting)
+  end
+
+  def get_module_relative_path(file_path, modulepath_directories)
+    parent_path = modulepath_directories.find { |path_dir| is_parent_dir_of(path_dir.downcase, file_path) }
+    return :no_path if parent_path.nil?
+
+    file_path.relative_path_from(Pathname.new(parent_path.downcase)).to_s.split(File::SEPARATOR)
+  end
+
+  def is_parent_dir_of(parent_dir, child_dir)
+    return false unless child_dir.to_s.start_with?(parent_dir)
+
+    parent_dir_path = Pathname.new(parent_dir)
+    child_dir.ascend do |up_dir|
+      return true if up_dir == parent_dir_path
+    end
+    return false
+  end
+
+  def dir_to_names(relative_path)
+    # Example definition dir: manifests in this path:
+    # <module name>/manifests/<module subdir>/<classfile>.pp
+    dir = relative_path[1]
+
+    # How can we get this result?
+    # If it is not an initial manifest, it must come from a module,
+    # and from the manifests dir there.  This may never get used...
+    return :bad_module_file unless %w[manifests functions types plans].include?(dir)
+    #return :bad_module_file unless dir == 'manifests' || dir == 'functions' || dir == 'types' || dir == 'plans'
```

Comment:
Let's use the explicit form (that's commented out) instead. The `include?` form creates an Array object and throws it away, which we've found to be more expensive and slower than the explicit check you had.